No module named '_zstd'

Hi,

I use pystow and never had any issue with it, thank your for maintaining it !
But I recently encountered the following issue in a CI environment for one of my projects, specifically in a cibuildwheel build check using CPython 3.14 on manylinux x86_64. I suspect that this environment doesn’t have zstd pre-installed.

```
import pystow
...
  ../venv/lib/python3.14/site-packages/pystow/__init__.py:3: in <module>
      from .api import (
  ../venv/lib/python3.14/site-packages/pystow/api.py:19: in <module>
      from .impl import Module, VersionHint
  ../venv/lib/python3.14/site-packages/pystow/impl.py:23: in <module>
      from . import utils
  ../venv/lib/python3.14/site-packages/pystow/utils/__init__.py:82: in <module>
      from .pydantic_utils import (
  ../venv/lib/python3.14/site-packages/pystow/utils/pydantic_utils.py:13: in <module>
      from .safe_open import (
  ../venv/lib/python3.14/site-packages/pystow/utils/safe_open.py:34: in <module>
      from compression import zstd
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
      """Python bindings to the Zstandard (zstd) compression library (RFC-8878)."""
      
      __all__ = (
          # compression.zstd
          'COMPRESSION_LEVEL_DEFAULT',
          'compress',
          'CompressionParameter',
          'decompress',
          'DecompressionParameter',
          'finalize_dict',
          'get_frame_info',
          'Strategy',
          'train_dict',
      
          # compression.zstd._zstdfile
          'open',
          'ZstdFile',
      
          # _zstd
          'get_frame_size',
          'zstd_version',
          'zstd_version_info',
          'ZstdCompressor',
          'ZstdDecompressor',
          'ZstdDict',
          'ZstdError',
      )
      
  >   import _zstd
  E   ModuleNotFoundError: No module named '_zstd'
  
  /opt/_internal/cpython-3.14.0/lib/python3.14/compression/zstd/__init__.py:29: ModuleNotFoundError
  =============================== warnings summary ===============================
```

A quick Google search led me to https://github.com/Rapptz/discord.py/discussions/10327#discussion-9001263 and a potential fix  https://github.com/aio-libs/aiohttp/pull/11611

Here is the CI setup info:

```
Current runner version: '2.335.1'
Runner Image Provisioner
  Hosted Compute Agent
  Version: 20260527.539
  Commit: a891dd388383b896fa6ac04a82c0b75cec981078
  Build Date: 2026-05-27T21:39:57Z
  Worker ID: {8a8bc72c-d3c8-4e96-9ac7-85d4cdacf5c7}
  Azure Region: westus
Operating System
  Ubuntu
  24.04.4
  LTS
Runner Image
  Image: ubuntu-24.04
  Version: 20260607.184.1
  Included Software: https://github.com/actions/runner-images/blob/ubuntu24/20260607.184/images/ubuntu/Ubuntu2404-Readme.md
  Image Release: https://github.com/actions/runner-images/releases/tag/ubuntu24%2F20260607.184
Download action repository 'actions/checkout@v2' (SHA:ee0669bd1cc54295c223e0bb666b733df41de1c5)
Download action repository 'pypa/cibuildwheel@v3.3.0' (SHA:63fd63b352a9a8bdcc24791c9dbee952ee9a8abc)
Download action repository 'actions/setup-python@v6' (SHA:a309ff8b426b58ec0e2a45f0f869d46889d02405)
```

Could you take a look ?

--- a/src/pystow/api.py
+++ b/src/pystow/api.py
@@ -6,7 +6,6 @@ import bz2
 import io
 import lzma
 import sqlite3
-import sys
 import typing
 from collections.abc import Generator, Mapping, Sequence
 from contextlib import contextmanager
@@ -18,11 +17,6 @@ from typing import TYPE_CHECKING, Any, Literal, overload
 from .constants import JSON, Provider
 from .impl import Module, VersionHint
 from .utils.download import DownloadKwargs
-
-if sys.version_info >= (3, 14):
-    from compression import zstd
-else:
-    from backports import zstd
 
 if TYPE_CHECKING:
     import bs4
@@ -710,7 +704,7 @@ def ensure_open_zstd(
     download_kwargs: DownloadKwargs | None = None,
     mode: Literal["r", "rb", "w", "wb", "rt", "wt"] = "rt",
     open_kwargs: Mapping[str, Any] | None = None,
-) -> Generator[zstd.ZstdFile, None, None]:
+) -> Generator[io.BufferedIOBase, None, None]:
     """Ensure a zstd-compressed file is downloaded and open a file inside it.
 
     :param key: The name of the module. No funny characters. The envvar `<key>_HOME`

--- a/src/pystow/impl.py
+++ b/src/pystow/impl.py
@@ -11,7 +11,6 @@ import lzma
 import os
 import pickle
 import sqlite3
-import sys
 import tarfile
 import typing
 from collections.abc import Callable, Generator, Mapping, Sequence
@@ -42,11 +41,7 @@ from .utils import (
     read_zipfile_csv,
 )
 from .utils.download import DownloadKwargs
-
-if sys.version_info >= (3, 14):
-    from compression import zstd
-else:
-    from backports import zstd
+from .utils.safe_open import zstd_open
 
 if TYPE_CHECKING:
     import botocore.client
@@ -750,7 +745,7 @@ class Module:
         download_kwargs: DownloadKwargs | None = None,
         mode: Literal["r", "rb", "w", "wb", "rt", "wt"] = "rt",
         open_kwargs: Mapping[str, Any] | None = None,
-    ) -> Generator[zstd.ZstdFile, None, None]:
+    ) -> Generator[io.BufferedIOBase, None, None]:
         """Ensure a zSTD-compressed file is downloaded and open a file inside it.
 
         :param subkeys: A sequence of additional strings to join. If none are given,
@@ -772,7 +767,7 @@ class Module:
         )
         open_kwargs = {} if open_kwargs is None else dict(open_kwargs)
         open_kwargs.setdefault("mode", mode)
-        with zstd.open(path, **open_kwargs) as file:
+        with zstd_open(path, **open_kwargs) as file:
             yield file
 
     # docstr-coverage:excused `overload`

--- a/src/pystow/utils/safe_open.py
+++ b/src/pystow/utils/safe_open.py
@@ -15,7 +15,9 @@ import urllib.request
 import zipfile
 from collections.abc import Generator, Mapping
 from pathlib import Path
-from typing import Any, BinaryIO, Literal, Never, TextIO, TypeGuard, cast, overload
+from typing import Any, BinaryIO, Literal, TextIO, TypeGuard, cast, overload
+
+from typing_extensions import Never
 
 from .io_typing import (
     _MODE_TO_SIMPLE,

--- a/src/pystow/utils/safe_open.py
+++ b/src/pystow/utils/safe_open.py
@@ -15,7 +15,7 @@ import urllib.request
 import zipfile
 from collections.abc import Generator, Mapping
 from pathlib import Path
-from typing import Any, BinaryIO, Literal, TextIO, TypeGuard, cast, overload
+from typing import Any, BinaryIO, Literal, Never, TextIO, TypeGuard, cast, overload
 
 from .io_typing import (
     _MODE_TO_SIMPLE,
@@ -30,10 +30,20 @@ from .io_typing import (
     ensure_sensible_newline,
 )
 
-if sys.version_info >= (3, 14):
-    from compression import zstd
-else:
-    from backports import zstd
+try:
+    if sys.version_info >= (3, 14):
+        from compression import zstd
+    else:
+        from backports import zstd
+
+    zstd_open = zstd.open
+
+except ImportError:
+
+    def zstd_open(*args: Any, **kwargs: Any) -> Never:
+        """Open a Zstandard compressed file in binary or text mode."""
+        raise RuntimeError("zstd is not available")
+
 
 __all__ = [
     "is_url",
@@ -47,6 +57,7 @@ __all__ = [
     "safe_write_text",
     "write_json",
     "write_yaml",
+    "zstd_open",
 ]
 
 
@@ -140,7 +151,7 @@ def safe_open(  # noqa:C901
                 with lzma.open(path, mode=mode, encoding=encoding, newline=newline) as file:
                     yield file  # type:ignore
             elif path.suffix.endswith(".zst"):
-                with zstd.open(path, mode=mode, encoding=encoding, newline=newline) as file:
+                with zstd_open(path, mode=mode, encoding=encoding, newline=newline) as file:
                     yield file  # type:ignore
             else:
                 with open(path, mode=mode, encoding=encoding, newline=newline) as file:

